csharp/src/Google.Protobuf/AsyncMessageExtensions.cs

```diff
@@ -0,0 +1,126 @@
+    /// </summary>
+    public static class AsyncMessageExtensions
+    {
+        /// <summary>
+        /// Merges data from the given byte array into an existing message.
+        /// </summary>
+        /// <param name="message">The message to merge the data into.</param>
+        /// <param name="data">The data to merge, which must be protobuf-encoded binary data.</param>
+        /// <param name="cancellationToken"></param>
+        public static async Task MergeFrom(this IAsyncMessage message, byte[] data, CancellationToken cancellationToken)
+        {
+            ProtoPreconditions.CheckNotNull(message, "message");
+            ProtoPreconditions.CheckNotNull(data, "data");
+            CodedInputStream input = new CodedInputStream(data);
+            await message.MergeFromAsync(input, cancellationToken).ConfigureAwait(false);
+            input.CheckReadEndOfStreamTag();
+        }
+
+        /// <summary>
+        /// Merges data from the given stream into an existing message.
+        /// </summary>
+        /// <param name="message">The message to merge the data into.</param>
+        /// <param name="input">Stream containing the data to merge, which must be protobuf-encoded binary data.</param>
+        /// <param name="cancellationToken"></param>
+        public static async Task MergeFromAsync(this IAsyncMessage message, Stream input, CancellationToken cancellationToken)
+        {
+            ProtoPreconditions.CheckNotNull(message, "message");
+            ProtoPreconditions.CheckNotNull(input, "input");
+            CodedInputStream codedInput = new CodedInputStream(input);
+            await message.MergeFromAsync(codedInput, cancellationToken).ConfigureAwait(false);
+            codedInput.CheckReadEndOfStreamTag();
+        }
+
+        /// <summary>
+        /// Merges length-delimited data from the given stream into an existing message.
+        /// </summary>
+        /// <remarks>
+        /// The stream is expected to contain a length and then the data. Only the amount of data
+        /// specified by the length will be consumed.
+        /// </remarks>
+        /// <param name="message">The message to merge the data into.</param>
+        /// <param name="input">Stream containing the data to merge, which must be protobuf-encoded binary data.</param>
+        /// <param name="cancellationToken"></param>
+        public static async Task MergeDelimitedFromAsync(this IAsyncMessage message, Stream input, CancellationToken cancellationToken)
+        {
+            ProtoPreconditions.CheckNotNull(message, "message");
+            ProtoPreconditions.CheckNotNull(input, "input");
+            int size = (int) await CodedInputStream.ReadRawVarint32Async(input, cancellationToken).ConfigureAwait(false);
+            Stream limitedStream = new LimitedInputStream(input, size);
+            await message.MergeFromAsync(limitedStream, cancellationToken);
```

Comment:
If you know of a checker to make sure that every asynchronous call has a cancellation token, that would be good too...